File: src/server/proxy/procListeners.ts
export interface ListeningPort {
  port: number;
  proto: 'tcp';
}

// Linux kernel TCP state constant (TCP_LISTEN). See net/tcp_states.h.
const LISTEN_STATE = '0A';

export function parseProcNetTcp(text: string): ListeningPort[] {
  const seen = new Set<number>();
  const out: ListeningPort[] = [];
  const lines = text.split('\n');
  for (const raw of lines) {
    const line = raw.trim();
    if (!line || line.startsWith('sl')) continue;
    // Format: "  N: <local_addr> <rem_addr> <st> ..."
    const parts = line.split(/\s+/);
    // After splitting, parts[0] is "N:", parts[1] is local_address, parts[2] rem, parts[3] state.
    if (parts.length < 4) continue;
    const local = parts[1];
    const state = parts[3];
    if (state !== LISTEN_STATE) continue; // not LISTEN
    const colon = local.lastIndexOf(':');
    if (colon < 0) continue;
    const portHex = local.slice(colon + 1);
    const port = parseInt(portHex, 16);
    if (!Number.isFinite(port) || port <= 0 || port > 65535) continue;
    if (seen.has(port)) continue;
    seen.add(port);
    out.push({ port, proto: 'tcp' });
  }
  out.sort((a, b) => a.port - b.port);
  return out;
}

File: server-elixir/lib/fbi/proxy/proc_listeners.ex
defmodule FBI.Proxy.ProcListeners do
  @moduledoc "Parses /proc/net/tcp output for LISTEN-state sockets."

  # Linux TCP_LISTEN constant.
  @listen_state "0A"

  @spec parse(binary()) :: [%{port: integer(), proto: :tcp}]
  def parse(text) when is_binary(text) do
    text
    |> String.split("\n", trim: false)
    |> Enum.reduce(%{seen: MapSet.new(), out: []}, &reduce_line/2)
    |> Map.fetch!(:out)
    |> Enum.reverse()
    |> Enum.sort_by(& &1.port)
  end

  defp reduce_line(raw, %{seen: seen, out: out} = acc) do
    line = String.trim(raw)

    cond do
      line == "" or String.starts_with?(line, "sl") ->
        acc

      true ->
        parts = String.split(line, ~r/\s+/, trim: true)

        case parts do
          [_idx, local, _rem, state | _] when state == @listen_state ->
            with [_addr, port_hex] <- String.split(local, ":", parts: 2),
                 {port, ""} <- Integer.parse(port_hex, 16),
                 true <- port > 0 and port <= 65535,
                 false <- MapSet.member?(seen, port) do
              %{seen: MapSet.put(seen, port), out: [%{port: port, proto: :tcp} | out]}
            else
              _ -> acc
            end

          _ ->
            acc
        end
    end
  end
end
